Add veterancy-based target filtering for weapons and warheads

Flactine opened the pull request:
This PR introduces **veterancy-based filtering** for both weapon targeting and warhead effect application.

---

## 1. Weapon target filtering by target veterancy

Weapons can now restrict which targets they are allowed to attack based on the target’s veterancy level.
This affects both weapon firing eligibility and weapon selection logic (except for special cases where weapon selection is fixed by the engine).

### New INI options:

```ini
[SOMEWEAPON]             ; WeaponType
CanTargetVeterancy=all     ; List of Affected Veterancy Enumeration (none|rookie|veteran|elite|all)
```

Targets whose veterancy is not included in the specified enumeration will not be considered valid targets for the weapon.

---

## 2. Warhead effect filtering by target veterancy

Warheads can now conditionally apply their effects based on the target’s veterancy level.
Unlike weapon filtering, this logic does **not** affect whether a weapon can fire, but instead determines whether the warhead’s effects will be applied after detonation.

### New INI options:

```ini
[SOMEWARHEAD]               ; WarheadType
AffectsVeterancy=all        ; List of Affected Veterancy Enumeration (none|rookie|veteran|elite|all)
```

* The warhead will only affect targets whose veterancy is explicitly listed in AffectsVeterancy.

---

---

---

本次更新新增了**基于目标等级（Veterancy）的筛选逻辑**，分别作用于**武器攻击判定**和**弹头效果生效判定**。

---

## 1. 基于目标等级的武器筛选

武器现在可以根据目标单位的经验等级，决定是否允许对其进行攻击。
该逻辑不仅影响武器能否开火，也会影响武器选择流程（部分引擎强制固定选择的特殊情况除外）。

### 新增 INI 选项（WeaponType）：

```ini
[SOMEWEAPON]             ; 武器类型
CanTargetVeterancy=all     ; List of Affected Veterancy Enumeration (none|rookie|veteran|elite|all)
```

当目标的等级未被列入指定枚举时，武器将不会将其视为可攻击目标。

---

## 2. 基于目标等级的弹头效果筛选

弹头现在可以根据目标单位的经验等级，决定爆炸后是否对其产生效果。
与武器筛选不同，此逻辑**不影响武器是否开火**，而是作用于爆炸结算阶段。

### 新增 INI 选项：

```ini
[SOMEWARHEAD]               ; 弹头类型
AffectsVeterancy=all        ; List of Affected Veterancy Enumeration (none|rookie|veteran|elite|all)
```

* 只有等级被列入指定枚举的目标才会受到弹头效果影响

Flactine committed:
Add veterancy threshold check for TechnoClass objects

Introduced IsVeterancyInThreshold to TechnoExt for checking if a TechnoClass object's veterancy level falls within a specified range. Updated relevant headers and added a corresponding method declaration in WarheadType's ExtData for future use.

Flactine committed:
Add veterancy threshold checks for weapons and warheads

Introduces veterancy-based threshold checks for WeaponType and WarheadType, allowing configuration of min/max veterancy for targeting and effect application. Updates relevant methods, serialization, and INI loading to support new AffectsAboveVeterancy, AffectsBelowVeterancy, CanTarget_MinVeterancy, and CanTarget_MaxVeterancy fields.

Flactine committed:
Add debug logs and use float suffixes for veterancy values

Added debug logging to IsVeterancyInThreshold for min, max, and veterancy values in TechnoExt. Updated veterancy-related default values in WarheadType and WeaponType headers to use explicit float suffixes for consistency.

Flactine committed:
Fix incorrect INI keys for veterancy percent loading

Corrected the code to read 'AffectsBelowVeterancy' and 'AffectsAboveVeterancy' into the appropriate variables instead of incorrectly reading them into percent variables.

Flactine committed:
Add veterancy threshold checks to targeting logic

Extended weapon and warhead targeting logic to include veterancy threshold checks alongside health checks. Updated relevant methods and hooks to ensure that actions such as firing, shrapnel targeting, and damage application now also consider the target's veterancy. Removed debug logging from IsVeterancyInThreshold for cleaner output. Adjusted INI loading to use float literals for veterancy thresholds.

Flactine committed:
Document veterancy-based targeting and warhead logic

Added documentation for new veterancy-based conditions: `AffectsAboveVeterancy` and `AffectsBelowVeterancy` for warhead detonation, and `CanTarget.MinVeterancy` and `CanTarget.MaxVeterancy` for weapon targeting. Clarified how veterancy values are interpreted and how non-trainable TechnoTypes are handled.

Flactine committed:
Update Whats-New.md

Flactine committed:
Merge branch 'develop' into CanTarget-MaxLevel

TaranDahl commented on docs/Whats-New.md:
```diff
@@ -465,6 +465,10 @@ New:
 - [Added Target Filtering Options to AttachEffect System](New-or-Enhanced-Logics.md#attached-effects) (by Flactine)
 - [Customize type selection for IFV](Fixed-or-Improved-Logics.md#customize-type-selection-for-ifv) (by NetsuNegi)
+- CellSpread in cylinder shape (by TaranDahl)
+- CellSpread damage check if victim is in air or on floor (by TaranDahl)
```
Wrong docs.

TaranDahl commented on src/Ext/WeaponType/Body.h:
```diff
@@ -114,6 +116,8 @@ class WeaponTypeExt
 			, CanTargetHouses { AffectedHouse::All }
 			, CanTarget_MaxHealth { 1.0 }
 			, CanTarget_MinHealth { 0.0 }
+			, CanTarget_MaxVeterancy { 2.0f }
```
Perhaps you should use Nullable and use `RulesClass::Instance->VeteranCap` as the default value?

Flactine committed:
Use dynamic veterancy cap for target filtering

Replaced hardcoded veterancy limits with RulesClass::Instance->VeteranCap in WarheadType and WeaponType target filtering logic. Updated documentation and credits to reflect the addition of veterancy-based target filtering.

Flactine committed:
Cast VeteranCap to float in WarheadType and WeaponType

Replaced direct assignment of RulesClass::Instance->VeteranCap with a static_cast to float in both WarheadType and WeaponType Body.h files to ensure type consistency.

Flactine committed:
Switch veterancy fields to Nullable<float> in WarheadType and WeaponType

Replaced Valueable<float> with Nullable<float> for veterancy-related fields in WarheadType and WeaponType headers. This allows these fields to represent unset/null values, improving flexibility in configuration.

Flactine committed:
Refactor veterancy threshold handling for warheads and weapons

Replaced Nullable with Valueable for AffectsAboveVeterancy and CanTarget_MinVeterancy to allow value resolution with VeteranCap. Updated related logic in WarheadTypeExt and WeaponTypeExt to use the new types and ensure correct threshold checks based on game rules.

Flactine committed:
Fix veterancy cap comparison logic in INI loading

Replaced hardcoded veterancy cap values with dynamic checks using RulesClass::Instance->VeteranCap in WarheadTypeExt and WeaponTypeExt INI loading. This ensures correct behavior when the veterancy cap is changed from its default value.

Flactine committed:
Docs

TaranDahl commented on src/Ext/WarheadType/Body.h:
```diff
@@ -192,6 +192,8 @@ class WarheadTypeExt
 		Valueable<double> AffectsBelowPercent;
 		Valueable<double> AffectsAbovePercent;
+		Nullable<float> AffectsBelowVeterancy;
+		Valueable<float> AffectsAboveVeterancy;
```
Why not use double?

TaranDahl commented on src/Ext/WeaponType/Body.cpp:
```diff
@@ -157,6 +164,7 @@ void WeaponTypeExt::ExtData::LoadFromINIFile(CCINIClass* const pINI)
 	// handle SkipWeaponPicking
 	if (this->CanTarget != AffectedTarget::All || this->CanTargetHouses != AffectedHouse::All
 		|| this->CanTarget_MaxHealth < 1.0 || this->CanTarget_MinHealth > 0.0
+		|| this->CanTarget_MaxVeterancy.Get(static_cast<float>(RulesClass::Instance->VeteranCap)) < static_cast<float>(RulesClass::Instance->VeteranCap) || this->CanTarget_MinVeterancy > 0.0f
```
You may test this in-game. I'm not sure if RulesClass::Instance has been loaded at this time.

Flactine replied:
It is loaded. Debug information can display "RulesClass::Instance->VeteranCap" at that time.

Flactine committed:
Refactor veterancy threshold types to double

Changed veterancy-related variables and function parameters from float to double in WarheadType and WeaponType extensions for improved precision and consistency. Updated related method calls and initializations accordingly.

Flactine committed:
Fix logic and type for veterancy and health checks

Corrected the parameter types for IsVeterancyInThreshold to use double instead of float. Fixed the logic for HealthCheck and VeterancyCheck assignments in WarheadTypeExt::ExtData::LoadFromINIFile to properly reflect the intended threshold checks.

TaranDahl commented on docs/Fixed-or-Improved-Logics.md:
```diff
@@ -2320,6 +2326,8 @@ In `rulesmd.ini`:
 [SOMEWARHEAD]               ; WarheadType
 AffectsBelowPercent=1.0     ; floating point value, percents or absolute
 AffectsAbovePercent=0.0     ; floating point value, percents or absolute
+AffectsBelowVeterancy=2.0     ; floating point value, percents or absolute, default to [General] -> VeteranCap
```
You forgot the docs.

TaranDahl commented on docs/New-or-Enhanced-Logics.md:
```diff
@@ -2856,6 +2862,8 @@ CanTarget=all            ; List of Affected Target Enumeration (none|land|water|
 CanTargetHouses=all      ; List of Affected House Enumeration (none|owner/self|allies/ally|team|enemies/enemy|all)
 CanTarget.MaxHealth=1.0  ; floating point value, percents or absolute
 CanTarget.MinHealth=0.0  ; floating point value, percents or absolute
+CanTarget.MaxVeterancy=2.0  ; floating point value, percents or absolute, default to [General] -> VeteranCap
```
Same

Flactine committed:
Clarify data types for Veterancy fields in docs

Updated documentation to specify that AffectsBelowVeterancy, AffectsAboveVeterancy, CanTarget.MaxVeterancy, and CanTarget.MinVeterancy are of type double, improving clarity for users configuring these values.

Flactine committed:
Clarify veterancy parameter documentation

Updated documentation for AffectsBelowVeterancy and CanTarget.MaxVeterancy to clarify that these parameters accept floating point values, percents, or absolute values, and to indicate their default behavior. This improves accuracy and consistency in the docs.

Flactine committed:
Fix alignment and spacing in logic documentation

Corrected indentation and spacing for configuration examples in Fixed-or-Improved-Logics.md and New-or-Enhanced-Logics.md to improve readability and consistency.

Coronia commented on docs/Whats-New.md:
```diff
@@ -465,6 +465,8 @@ New:
 - [Added Target Filtering Options to AttachEffect System](New-or-Enhanced-Logics.md#attached-effects) (by Flactine)
 - [Customize type selection for IFV](Fixed-or-Improved-Logics.md#customize-type-selection-for-ifv) (by NetsuNegi)
+- [Weapon target filtering by target veterancy](New-or-Enhanced-Logics.md#weapon-targeting-filter) (by Flactine)
+- [Warhead effect filtering by target veterancy](Fixed-or-Improved-Logics.md#customizable-warhead-trigger-conditions) (by Flactine)
```
these 2 lines should be put at the end of this section

Coronia reviewed (approved):
maybe it's better to put veterancy in an enum, so it can be represented in a clearer way than having to remember what 0.0, 1.0 and 2.0 stand for, and could be reused in the future

ZivDero commented on CREDITS.md:
```diff
@@ -656,7 +656,9 @@ This page lists all the individual contributions to the project by their author.
-- **Flactine** - add target filtering options to attacheffect system
+- **Flactine** 
+  - add target filtering options to attacheffect system
+  - add veterancy-based target filtering for weapons and warheads
```
Capitalization please.

Metadorius commented on src/Ext/Techno/Body.cpp:
```diff
@@ -743,6 +743,16 @@ bool TechnoExt::IsHealthInThreshold(TechnoClass* pObject, double min, double max
+	if (pTechno->GetTechnoType()->Trainable)
+		veterancy = pTechno->Veterancy.Veterancy;
+
+	return veterancy >= min && veterancy <= max;
```
One of the limits should be inclusive, other exclusive. Otherwise imagine if you make one warhead effect that works on 0-1, another that works on 1-2. With your implementation it's impossible to create non-overlapping ranges of effect in a good fashion. See how `IsHealthInThreshold` handles this.

I think the lower bound should be inclusive, and the upper bound should be exclusive. The reasoning is that the game itself uses it like this. [0, 1) is Rookie, [1, 2) is Veteran, and [2, +inf) is Elite. This means


```suggestion
	return veterancy >= min && veterancy < max;
```

should be used.

However, the same issue as with `IsHealthInThreshold` arises when you can't then cover all ranges in a nice way, hence why we made 0 HP follow a slightly different set of rules. Accepting ideas on how to handle this similarly.

ZivDero reviewed (changes requested):
It looks solid overall, but I can't help shake the feeling that the "can target" and "can affect" stuff is getting a bit spaghetti. Could combine it all into single "can target" and "can affect" functions (with subfunctions for the various logics). Though this is not the fault of this specific PR.

Metadorius commented on src/Ext/WeaponType/Body.cpp:
```diff
@@ -46,6 +46,11 @@ bool WeaponTypeExt::ExtData::IsHealthInThreshold(TechnoClass* pTarget) const
 
+bool WeaponTypeExt::ExtData::IsVeterancyInThreshold(TechnoClass* pTarget) const
+{
+	return TechnoExt::IsVeterancyInThreshold(pTarget, this->CanTarget_MinVeterancy, this->CanTarget_MaxVeterancy.Get(RulesClass::Instance->VeteranCap));
```
> `this->CanTarget_MaxVeterancy.Get(RulesClass::Instance->VeteranCap)`

@ZivDero or whoever, defaultable template pleeeeeeeeeeease 🥺

Flactine replied:
Is it OK to change into "RulesClass::Instance->VeteranCap + 1.0"?

Metadorius replied:
maybe... not sure about the usability

@Starkku any ideas?

Flactine committed:
Update src/Ext/Techno/Body.cpp

Co-authored-by: Kerbiter <crabiter@vivaldi.net>

Flactine committed:
Update CREDITS.md

Flactine committed:
Update Whats-New.md

Flactine committed:
Adjust veterancy threshold logic for warheads and weapons

Updated the default upper bound for veterancy checks to use VeteranCap + 1.0 instead of VeteranCap, ensuring that the threshold is exclusive and matches documentation. Updated related documentation and developer warnings to reflect this change.

TaranDahl commented:
> maybe it's better to put veterancy in an enum, so it can be represented in a clearer way than having to remember what 0.0, 1.0 and 2.0 stand for, and could be reused in the future

@Flactine You can make that flag read both the enum (Rookie, Veteran, Elite) and float types simultaneously.
After that, I think this PR can be merged? @Coronia @ZivDero @Metadorius

NetsuNegi committed:
Merge branch 'develop' into CanTarget-MaxLevel

Flactine committed:
Value to Enum

TaranDahl commented:
@Coronia @Metadorius Going to merge this in 24h if no one has further instructions.
